bpf/lib/l3.h

```diff
@@ -126,7 +127,8 @@ static __always_inline int ipv4_local_delivery(struct __ctx_buff *ctx, int l3_of
 
 #if defined(USE_BPF_PROG_FOR_INGRESS_POLICY) && \
 	!defined(FORCE_LOCAL_POLICY_EVAL_AT_SOURCE)
-	ctx->mark = (seclabel << 16) | MARK_MAGIC_IDENTITY;
+	ctk->mark |= MARK_MAGIC_IDENTITY;
+	set_identity_mark(ctx, seclabel)
```

Comment:
So, I don't understand the details but from reading the commit message my understanding is that the original code zeroed some needed bits by making the assignment. It's not clear what the bits that we need are, but I wonder: do we need _all_ the bits from the original mark or do we need to mask some of them out and only keep the ones we need?

Reply:
Thanks for the review! Great question.

In multi-cluster environments, the identity is expanded to 24 bits: 8 bits cluster-id + 16bit identity within that cluster. When importing identities from remote clusters, we prepend the cluster-id of the source cluster into any identities for use in the local cluster (eg inserting into ipcache, etc.).

Prior to this PR we would only set the upper 16 bits in the mark which denote the locally-significant identity. And as you point out, we would zero most of the rest of the bits, including the lowest 8 bits range which is supposed to indicate which cluster that the identity is significant for. This means that instead of treating it as an identity from a remote cluster, we treat it as an identity from the local cluster for the same 16-bit number (which is likely not consistent between clusters as they allocate numeric identities independently). In most cases this basically means that we would misidentify this traffic as some identity that doesn't exist and drop it. There's worse possible cases, but in practice it doesn't make sense to configure this particular configuration in the datapath (direct-routing + tunnel) so this PR is really just trying to ensure that if and when this path is properly enabled, there isn't some weird identity misuse that could cause havoc in the cluster.

We also re-use the `set_identity_mark()` helper which is just generally good for more consistent code, easier readability, making sure that we can find all locations for a particular use case, etc.

Reply:
Thanks for the detailed reply!